### PR TITLE
Bluetooth: host: gatt: Properly check subscription status

### PR DESCRIPTION
**Pull request by jori-nordic**
We had two bugs:

- we didn't check the subscription, only the fact that the passed chrc can
be subscribed to.

- that check only worked if we passed the chrc descriptor handle, not its
value handle. It seems it can lead to a nullptr dereference.

This also cherry-picks a babblesim test originally made by @laxiLang . 

Using that test, it is very easy to reproduce the issue, revert the fix commit, run the test and observe it failing when passing a characteristic value instead of a descriptor.

### CODE REVIEW

**Review by Thalley — CHANGES_REQUESTED** [review]
I wonder if it wouldn't have been easier to expand the existing `test_notify` functions to also do multiple, rather than recreating the entire test setup again :)

**Comment by jori-nordic** [comment]
> I wonder if it wouldn't have been easier to expand the existing `test_notify` functions to also do multiple, rather than recreating the entire test setup again :)

yeah but @hermabe is about to merge a huge rework of the existing notify test, and I think it's trying to test a lot of different things now. I can always merge it later.

File: tests/bluetooth/bsim_bt/bsim_test_notify_multiple/src/gatt_client_test.c [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+	UNSET_FLAG(flag_subscribed);
+	sub_params_long.value_handle = long_chrc_handle;
+	err = bt_gatt_unsubscribe(g_conn, &sub_params_long);
+	if (err < 0) {
+		FAIL("Failed to unsubscribe\n");
+	} else {
+		printk("Unsubscribe request sent\n");
+	}
+	WAIT_FOR_FLAG(flag_subscribed);
+}
```
[the remark this answers is not in the document]
Comment (jori-nordic) [reply tests/bluetooth/bsim_bt/bsim_test_notify_multiple/src/gatt_client_test.c:343]: yeah I tried to do that initially and using the same params, but there's a check in gatt.c that params is not the same address as a previous sub. so I gave up.
I'll apply your suggestion.

**Comment by hermabe** [comment]
> > I wonder if it wouldn't have been easier to expand the existing `test_notify` functions to also do multiple, rather than recreating the entire test setup again :)
> 
> yeah but @hermabe is about to merge a huge rework of the existing notify test, and I think it's trying to test a lot of different things now. I can always merge the two suites later.

I think we should create a library with common operations for bsim tests that we can just link to. Many tests are currently doing very similar things to advertise, scan, connect and set up database. In general the bsim setup have a lot of stuff copy-pasted. A library with common operations and the pytest plugin would make things much nicer.

**Comment by jori-nordic** [comment]
> > > I wonder if it wouldn't have been easier to expand the existing `test_notify` functions to also do multiple, rather than recreating the entire test setup again :)
> > 
> > 
> > yeah but @hermabe is about to merge a huge rework of the existing notify test, and I think it's trying to test a lot of different things now. I can always merge the two suites later.
> 
> I think we should create a library with common operations for bsim tests that we can just link to. Many tests are currently doing very similar things to advertise, scan, connect and set up database. In general the bsim setup have a lot of stuff copy-pasted. A library with common operations and the pytest plugin would make things much nicer.

wholeheartedly agree

**Comment by Thalley** [comment]
> > I think we should create a library with common operations for bsim tests that we can just link to. Many tests are currently doing very similar things to advertise, scan, connect and set up database. In general the bsim setup have a lot of stuff copy-pasted. A library with common operations and the pytest plugin would make things much nicer.
> 
> wholeheartedly agree

Ditto. We just need to keep in mind that BSIM isn't specific to Bluetooth (although Bluetooth may be the only use of it in Zephyr so far), but we should/could definitely make some common functionality that can be used by all the Bluetooth tests.

File: tests/bluetooth/bsim_bt/bsim_test_notify_multiple/src/gatt_client_test.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (jori-nordic) [reply tests/bluetooth/bsim_bt/bsim_test_notify_multiple/src/gatt_client_test.c]: went one level more. tell me if you'd still prefer to separate the sub and unsub fns

**Review by Thalley — CHANGES_REQUESTED** [review]

File: tests/bluetooth/bsim_bt/bsim_test_notify_multiple/src/gatt_client_test.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (Thalley) [reply tests/bluetooth/bsim_bt/bsim_test_notify_multiple/src/gatt_client_test.c]: See https://github.com/zephyrproject-rtos/zephyr/pull/49211#discussion_r949190546

File: tests/bluetooth/bsim_bt/bsim_test_notify_multiple/src/gatt_client_test.c [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+	}
+
+	if (err < 0) {
+		FAIL("Failed to %ssubscribe (err %d)\n", subscribe ? "un":"", err);
+	} else {
+		printk("%ssubscribe request sent\n", subscribe ? "un":"");
+	}
+
+	WAIT_FOR_FLAG(flag_subscribed);
+}
```
[the remark this answers is not in the document]
Comment (jori-nordic) [reply tests/bluetooth/bsim_bt/bsim_test_notify_multiple/src/gatt_client_test.c]: agreed

**Review by Thalley — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**